src/components/modals/ContactsPermissionModal.tsx

```diff
@@ -16,15 +22,52 @@ interface Props {
   bridge: AirshipBridge<ContactsPermissionResult | undefined>
 }
 
+export let showModalPromise: Promise<ContactsPermissionResult | undefined>
+
+/**
+ * Shows one instance of this modal asking if the user wants to grant contacts
+ * permissions, if the system contacts permission isn't granted.
+ *
+ * @returns Result of Contacts Access modal or undefined if no modal shown or
+ * the modal was dismissed somehow.
+ */
+export async function maybeShowContactsPermissionModal(): Promise<ContactsPermissionResult | undefined> {
+  const contactsPermissionOn = (await check(permissionNames.contacts).catch(_error => 'denied')) === 'granted'
+  if (contactsPermissionOn || showModalPromise != null) return undefined
+
+  showModalPromise = (async () => {
+    return await Airship.show<ContactsPermissionResult | undefined>(bridge => <ContactsPermissionModal bridge={bridge} />)
+  })()
+
+  return await showModalPromise
+}
```

Comment:
I like this new function, but the logic is a little off. Try this:

```ts
export function maybeShowContactsPermissionModal() : ThunkAction<Promise<ContactsPermissionResult | undefined>> {
  return async (dispatch, getState) => {
    const state = getState()
    const {account} = state.core

    // Bail if we are currently showing the modal:
    if (showModalPromise != null) return

    // Bail if we have ever show the modal:
    const {contactsPermissionShown} = state.ui.settings.contactsPermissionShown
    if (contactsPermissionShown) return

    // Bail if we already have permission:
   const contactsPermissionOn = (await check(permissionNames.contacts).catch(_error => 'denied')) === 'granted'
  if (contactsPermissionOn) return

     // Show the modal:
    const showModalPromise = Airship.show<ContactsPermissionResult | undefined>(bridge => <ContactsPermissionModal bridge={bridge} />)
    await writeContactsPermissionShown(account, true)

    return await showModalPromise
  }
}
```

Changes:

- Turned this into a thunk, so we can access redux.
- Broke out & commented the reasons why the modal wouldn't show
- Got rid of the IIFE around `Airship.show`
- Writing to disk, since we can handle that out here now.